// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: homologous_range.proto

package com.verily.genomewarp;

public final class HomologousRangeOuterClass {
  private HomologousRangeOuterClass() {}
  public static void registerAllExtensions(
      com.google.protobuf.ExtensionRegistry registry) {
  }
  public interface HomologousRangeOrBuilder extends
      // @@protoc_insertion_point(interface_extends:com.verily.genomewarp.HomologousRange)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>optional .google.genomics.v1.Range query_range = 1;</code>
     *
     * <pre>
     * The range of the region in the query genome assembly.
     * </pre>
     */
    boolean hasQueryRange();
    /**
     * <code>optional .google.genomics.v1.Range query_range = 1;</code>
     *
     * <pre>
     * The range of the region in the query genome assembly.
     * </pre>
     */
    com.google.genomics.v1.Range getQueryRange();
    /**
     * <code>optional .google.genomics.v1.Range query_range = 1;</code>
     *
     * <pre>
     * The range of the region in the query genome assembly.
     * </pre>
     */
    com.google.genomics.v1.RangeOrBuilder getQueryRangeOrBuilder();

    /**
     * <code>optional .google.genomics.v1.Range target_range = 2;</code>
     *
     * <pre>
     * The range of the region in the target genome assembly.
     * </pre>
     */
    boolean hasTargetRange();
    /**
     * <code>optional .google.genomics.v1.Range target_range = 2;</code>
     *
     * <pre>
     * The range of the region in the target genome assembly.
     * </pre>
     */
    com.google.genomics.v1.Range getTargetRange();
    /**
     * <code>optional .google.genomics.v1.Range target_range = 2;</code>
     *
     * <pre>
     * The range of the region in the target genome assembly.
     * </pre>
     */
    com.google.genomics.v1.RangeOrBuilder getTargetRangeOrBuilder();

    /**
     * <code>optional .com.verily.genomewarp.HomologousRange.TargetStrand target_strand = 3;</code>
     */
    int getTargetStrandValue();
    /**
     * <code>optional .com.verily.genomewarp.HomologousRange.TargetStrand target_strand = 3;</code>
     */
    com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand getTargetStrand();

    /**
     * <code>optional .com.verily.genomewarp.HomologousRange.RegionType region_type = 4;</code>
     */
    int getRegionTypeValue();
    /**
     * <code>optional .com.verily.genomewarp.HomologousRange.RegionType region_type = 4;</code>
     */
    com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType getRegionType();
  }
  /**
   * Protobuf type {@code com.verily.genomewarp.HomologousRange}
   *
   * <pre>
   * Message that links a genomic range within a query genome assembly to its
   * corresponding region in a target genome assembly. The structure is required
   * for transformation of gVCF or VCF+confident region BED files from one
   * assembly to another.
   * Next ID: 5
   * </pre>
   */
  public  static final class HomologousRange extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:com.verily.genomewarp.HomologousRange)
      HomologousRangeOrBuilder {
    // Use HomologousRange.newBuilder() to construct.
    private HomologousRange(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private HomologousRange() {
      targetStrand_ = 0;
      regionType_ = 0;
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private HomologousRange(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 10: {
              com.google.genomics.v1.Range.Builder subBuilder = null;
              if (queryRange_ != null) {
                subBuilder = queryRange_.toBuilder();
              }
              queryRange_ = input.readMessage(com.google.genomics.v1.Range.parser(), extensionRegistry);
              if (subBuilder != null) {
                subBuilder.mergeFrom(queryRange_);
                queryRange_ = subBuilder.buildPartial();
              }

              break;
            }
            case 18: {
              com.google.genomics.v1.Range.Builder subBuilder = null;
              if (targetRange_ != null) {
                subBuilder = targetRange_.toBuilder();
              }
              targetRange_ = input.readMessage(com.google.genomics.v1.Range.parser(), extensionRegistry);
              if (subBuilder != null) {
                subBuilder.mergeFrom(targetRange_);
                targetRange_ = subBuilder.buildPartial();
              }

              break;
            }
            case 24: {
              int rawValue = input.readEnum();

              targetStrand_ = rawValue;
              break;
            }
            case 32: {
              int rawValue = input.readEnum();

              regionType_ = rawValue;
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.verily.genomewarp.HomologousRangeOuterClass.internal_static_com_verily_genomewarp_HomologousRange_descriptor;
    }

    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.verily.genomewarp.HomologousRangeOuterClass.internal_static_com_verily_genomewarp_HomologousRange_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.class, com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.Builder.class);
    }

    /**
     * Protobuf enum {@code com.verily.genomewarp.HomologousRange.TargetStrand}
     *
     * <pre>
     * The genome strand on which the target range aligns. By convention, the
     * query assembly is always on the positive strand. This value is required
     * for reverse complement and general alignment-based coordinate
     * transformations.
     * </pre>
     */
    public enum TargetStrand
        implements com.google.protobuf.ProtocolMessageEnum {
      /**
       * <code>UNKNOWN_TARGET_STRAND = 0;</code>
       */
      UNKNOWN_TARGET_STRAND(0, 0),
      /**
       * <code>POSITIVE_STRAND = 1;</code>
       */
      POSITIVE_STRAND(1, 1),
      /**
       * <code>NEGATIVE_STRAND = 2;</code>
       */
      NEGATIVE_STRAND(2, 2),
      UNRECOGNIZED(-1, -1),
      ;

      /**
       * <code>UNKNOWN_TARGET_STRAND = 0;</code>
       */
      public static final int UNKNOWN_TARGET_STRAND_VALUE = 0;
      /**
       * <code>POSITIVE_STRAND = 1;</code>
       */
      public static final int POSITIVE_STRAND_VALUE = 1;
      /**
       * <code>NEGATIVE_STRAND = 2;</code>
       */
      public static final int NEGATIVE_STRAND_VALUE = 2;


      public final int getNumber() {
        if (index == -1) {
          throw new java.lang.IllegalArgumentException(
              "Can't get the number of an unknown enum value.");
        }
        return value;
      }

      public static TargetStrand valueOf(int value) {
        switch (value) {
          case 0: return UNKNOWN_TARGET_STRAND;
          case 1: return POSITIVE_STRAND;
          case 2: return NEGATIVE_STRAND;
          default: return null;
        }
      }

      public static com.google.protobuf.Internal.EnumLiteMap<TargetStrand>
          internalGetValueMap() {
        return internalValueMap;
      }
      private static final com.google.protobuf.Internal.EnumLiteMap<
          TargetStrand> internalValueMap =
            new com.google.protobuf.Internal.EnumLiteMap<TargetStrand>() {
              public TargetStrand findValueByNumber(int number) {
                return TargetStrand.valueOf(number);
              }
            };

      public final com.google.protobuf.Descriptors.EnumValueDescriptor
          getValueDescriptor() {
        return getDescriptor().getValues().get(index);
      }
      public final com.google.protobuf.Descriptors.EnumDescriptor
          getDescriptorForType() {
        return getDescriptor();
      }
      public static final com.google.protobuf.Descriptors.EnumDescriptor
          getDescriptor() {
        return com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.getDescriptor().getEnumTypes().get(0);
      }

      private static final TargetStrand[] VALUES = values();

      public static TargetStrand valueOf(
          com.google.protobuf.Descriptors.EnumValueDescriptor desc) {
        if (desc.getType() != getDescriptor()) {
          throw new java.lang.IllegalArgumentException(
            "EnumValueDescriptor is not for this type.");
        }
        if (desc.getIndex() == -1) {
          return UNRECOGNIZED;
        }
        return VALUES[desc.getIndex()];
      }

      private final int index;
      private final int value;

      private TargetStrand(int index, int value) {
        this.index = index;
        this.value = value;
      }

      // @@protoc_insertion_point(enum_scope:com.verily.genomewarp.HomologousRange.TargetStrand)
    }

    /**
     * Protobuf enum {@code com.verily.genomewarp.HomologousRange.RegionType}
     *
     * <pre>
     * The type of the region, indicating the complexity required to perform
     * the variant coordinate transform.
     * </pre>
     */
    public enum RegionType
        implements com.google.protobuf.ProtocolMessageEnum {
      /**
       * <code>UNKNOWN_REGION_TYPE = 0;</code>
       */
      UNKNOWN_REGION_TYPE(0, 0),
      /**
       * <code>IDENTICAL = 1;</code>
       *
       * <pre>
       * Sequences are identical between genomes
       * </pre>
       */
      IDENTICAL(1, 1),
      /**
       * <code>MISMATCHED_BASES = 4;</code>
       *
       * <pre>
       * This has mismatches but no indels
       * </pre>
       */
      MISMATCHED_BASES(2, 4),
      /**
       * <code>ALIGNMENT_REQUIRED = 3;</code>
       *
       * <pre>
       * This requires full alignment
       * </pre>
       */
      ALIGNMENT_REQUIRED(3, 3),
      UNRECOGNIZED(-1, -1),
      ;

      /**
       * <code>UNKNOWN_REGION_TYPE = 0;</code>
       */
      public static final int UNKNOWN_REGION_TYPE_VALUE = 0;
      /**
       * <code>IDENTICAL = 1;</code>
       *
       * <pre>
       * Sequences are identical between genomes
       * </pre>
       */
      public static final int IDENTICAL_VALUE = 1;
      /**
       * <code>MISMATCHED_BASES = 4;</code>
       *
       * <pre>
       * This has mismatches but no indels
       * </pre>
       */
      public static final int MISMATCHED_BASES_VALUE = 4;
      /**
       * <code>ALIGNMENT_REQUIRED = 3;</code>
       *
       * <pre>
       * This requires full alignment
       * </pre>
       */
      public static final int ALIGNMENT_REQUIRED_VALUE = 3;


      public final int getNumber() {
        if (index == -1) {
          throw new java.lang.IllegalArgumentException(
              "Can't get the number of an unknown enum value.");
        }
        return value;
      }

      public static RegionType valueOf(int value) {
        switch (value) {
          case 0: return UNKNOWN_REGION_TYPE;
          case 1: return IDENTICAL;
          case 4: return MISMATCHED_BASES;
          case 3: return ALIGNMENT_REQUIRED;
          default: return null;
        }
      }

      public static com.google.protobuf.Internal.EnumLiteMap<RegionType>
          internalGetValueMap() {
        return internalValueMap;
      }
      private static final com.google.protobuf.Internal.EnumLiteMap<
          RegionType> internalValueMap =
            new com.google.protobuf.Internal.EnumLiteMap<RegionType>() {
              public RegionType findValueByNumber(int number) {
                return RegionType.valueOf(number);
              }
            };

      public final com.google.protobuf.Descriptors.EnumValueDescriptor
          getValueDescriptor() {
        return getDescriptor().getValues().get(index);
      }
      public final com.google.protobuf.Descriptors.EnumDescriptor
          getDescriptorForType() {
        return getDescriptor();
      }
      public static final com.google.protobuf.Descriptors.EnumDescriptor
          getDescriptor() {
        return com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.getDescriptor().getEnumTypes().get(1);
      }

      private static final RegionType[] VALUES = values();

      public static RegionType valueOf(
          com.google.protobuf.Descriptors.EnumValueDescriptor desc) {
        if (desc.getType() != getDescriptor()) {
          throw new java.lang.IllegalArgumentException(
            "EnumValueDescriptor is not for this type.");
        }
        if (desc.getIndex() == -1) {
          return UNRECOGNIZED;
        }
        return VALUES[desc.getIndex()];
      }

      private final int index;
      private final int value;

      private RegionType(int index, int value) {
        this.index = index;
        this.value = value;
      }

      // @@protoc_insertion_point(enum_scope:com.verily.genomewarp.HomologousRange.RegionType)
    }

    public static final int QUERY_RANGE_FIELD_NUMBER = 1;
    private com.google.genomics.v1.Range queryRange_;
    /**
     * <code>optional .google.genomics.v1.Range query_range = 1;</code>
     *
     * <pre>
     * The range of the region in the query genome assembly.
     * </pre>
     */
    public boolean hasQueryRange() {
      return queryRange_ != null;
    }
    /**
     * <code>optional .google.genomics.v1.Range query_range = 1;</code>
     *
     * <pre>
     * The range of the region in the query genome assembly.
     * </pre>
     */
    public com.google.genomics.v1.Range getQueryRange() {
      return queryRange_ == null ? com.google.genomics.v1.Range.getDefaultInstance() : queryRange_;
    }
    /**
     * <code>optional .google.genomics.v1.Range query_range = 1;</code>
     *
     * <pre>
     * The range of the region in the query genome assembly.
     * </pre>
     */
    public com.google.genomics.v1.RangeOrBuilder getQueryRangeOrBuilder() {
      return getQueryRange();
    }

    public static final int TARGET_RANGE_FIELD_NUMBER = 2;
    private com.google.genomics.v1.Range targetRange_;
    /**
     * <code>optional .google.genomics.v1.Range target_range = 2;</code>
     *
     * <pre>
     * The range of the region in the target genome assembly.
     * </pre>
     */
    public boolean hasTargetRange() {
      return targetRange_ != null;
    }
    /**
     * <code>optional .google.genomics.v1.Range target_range = 2;</code>
     *
     * <pre>
     * The range of the region in the target genome assembly.
     * </pre>
     */
    public com.google.genomics.v1.Range getTargetRange() {
      return targetRange_ == null ? com.google.genomics.v1.Range.getDefaultInstance() : targetRange_;
    }
    /**
     * <code>optional .google.genomics.v1.Range target_range = 2;</code>
     *
     * <pre>
     * The range of the region in the target genome assembly.
     * </pre>
     */
    public com.google.genomics.v1.RangeOrBuilder getTargetRangeOrBuilder() {
      return getTargetRange();
    }

    public static final int TARGET_STRAND_FIELD_NUMBER = 3;
    private int targetStrand_;
    /**
     * <code>optional .com.verily.genomewarp.HomologousRange.TargetStrand target_strand = 3;</code>
     */
    public int getTargetStrandValue() {
      return targetStrand_;
    }
    /**
     * <code>optional .com.verily.genomewarp.HomologousRange.TargetStrand target_strand = 3;</code>
     */
    public com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand getTargetStrand() {
      com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand result = com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand.valueOf(targetStrand_);
      return result == null ? com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand.UNRECOGNIZED : result;
    }

    public static final int REGION_TYPE_FIELD_NUMBER = 4;
    private int regionType_;
    /**
     * <code>optional .com.verily.genomewarp.HomologousRange.RegionType region_type = 4;</code>
     */
    public int getRegionTypeValue() {
      return regionType_;
    }
    /**
     * <code>optional .com.verily.genomewarp.HomologousRange.RegionType region_type = 4;</code>
     */
    public com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType getRegionType() {
      com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType result = com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType.valueOf(regionType_);
      return result == null ? com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType.UNRECOGNIZED : result;
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      if (queryRange_ != null) {
        output.writeMessage(1, getQueryRange());
      }
      if (targetRange_ != null) {
        output.writeMessage(2, getTargetRange());
      }
      if (targetStrand_ != com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand.UNKNOWN_TARGET_STRAND.getNumber()) {
        output.writeEnum(3, targetStrand_);
      }
      if (regionType_ != com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType.UNKNOWN_REGION_TYPE.getNumber()) {
        output.writeEnum(4, regionType_);
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (queryRange_ != null) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(1, getQueryRange());
      }
      if (targetRange_ != null) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(2, getTargetRange());
      }
      if (targetStrand_ != com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand.UNKNOWN_TARGET_STRAND.getNumber()) {
        size += com.google.protobuf.CodedOutputStream
          .computeEnumSize(3, targetStrand_);
      }
      if (regionType_ != com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType.UNKNOWN_REGION_TYPE.getNumber()) {
        size += com.google.protobuf.CodedOutputStream
          .computeEnumSize(4, regionType_);
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code com.verily.genomewarp.HomologousRange}
     *
     * <pre>
     * Message that links a genomic range within a query genome assembly to its
     * corresponding region in a target genome assembly. The structure is required
     * for transformation of gVCF or VCF+confident region BED files from one
     * assembly to another.
     * Next ID: 5
     * </pre>
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:com.verily.genomewarp.HomologousRange)
        com.verily.genomewarp.HomologousRangeOuterClass.HomologousRangeOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return com.verily.genomewarp.HomologousRangeOuterClass.internal_static_com_verily_genomewarp_HomologousRange_descriptor;
      }

      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return com.verily.genomewarp.HomologousRangeOuterClass.internal_static_com_verily_genomewarp_HomologousRange_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.class, com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.Builder.class);
      }

      // Construct using com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
        }
      }
      public Builder clear() {
        super.clear();
        if (queryRangeBuilder_ == null) {
          queryRange_ = null;
        } else {
          queryRange_ = null;
          queryRangeBuilder_ = null;
        }
        if (targetRangeBuilder_ == null) {
          targetRange_ = null;
        } else {
          targetRange_ = null;
          targetRangeBuilder_ = null;
        }
        targetStrand_ = 0;

        regionType_ = 0;

        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return com.verily.genomewarp.HomologousRangeOuterClass.internal_static_com_verily_genomewarp_HomologousRange_descriptor;
      }

      public com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange getDefaultInstanceForType() {
        return com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.getDefaultInstance();
      }

      public com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange build() {
        com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange buildPartial() {
        com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange result = new com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange(this);
        if (queryRangeBuilder_ == null) {
          result.queryRange_ = queryRange_;
        } else {
          result.queryRange_ = queryRangeBuilder_.build();
        }
        if (targetRangeBuilder_ == null) {
          result.targetRange_ = targetRange_;
        } else {
          result.targetRange_ = targetRangeBuilder_.build();
        }
        result.targetStrand_ = targetStrand_;
        result.regionType_ = regionType_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange) {
          return mergeFrom((com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange other) {
        if (other == com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.getDefaultInstance()) return this;
        if (other.hasQueryRange()) {
          mergeQueryRange(other.getQueryRange());
        }
        if (other.hasTargetRange()) {
          mergeTargetRange(other.getTargetRange());
        }
        if (other.targetStrand_ != 0) {
          setTargetStrandValue(other.getTargetStrandValue());
        }
        if (other.regionType_ != 0) {
          setRegionTypeValue(other.getRegionTypeValue());
        }
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }

      private com.google.genomics.v1.Range queryRange_ = null;
      private com.google.protobuf.SingleFieldBuilder<
          com.google.genomics.v1.Range, com.google.genomics.v1.Range.Builder, com.google.genomics.v1.RangeOrBuilder> queryRangeBuilder_;
      /**
       * <code>optional .google.genomics.v1.Range query_range = 1;</code>
       *
       * <pre>
       * The range of the region in the query genome assembly.
       * </pre>
       */
      public boolean hasQueryRange() {
        return queryRangeBuilder_ != null || queryRange_ != null;
      }
      /**
       * <code>optional .google.genomics.v1.Range query_range = 1;</code>
       *
       * <pre>
       * The range of the region in the query genome assembly.
       * </pre>
       */
      public com.google.genomics.v1.Range getQueryRange() {
        if (queryRangeBuilder_ == null) {
          return queryRange_ == null ? com.google.genomics.v1.Range.getDefaultInstance() : queryRange_;
        } else {
          return queryRangeBuilder_.getMessage();
        }
      }
      /**
       * <code>optional .google.genomics.v1.Range query_range = 1;</code>
       *
       * <pre>
       * The range of the region in the query genome assembly.
       * </pre>
       */
      public Builder setQueryRange(com.google.genomics.v1.Range value) {
        if (queryRangeBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          queryRange_ = value;
          onChanged();
        } else {
          queryRangeBuilder_.setMessage(value);
        }

        return this;
      }
      /**
       * <code>optional .google.genomics.v1.Range query_range = 1;</code>
       *
       * <pre>
       * The range of the region in the query genome assembly.
       * </pre>
       */
      public Builder setQueryRange(
          com.google.genomics.v1.Range.Builder builderForValue) {
        if (queryRangeBuilder_ == null) {
          queryRange_ = builderForValue.build();
          onChanged();
        } else {
          queryRangeBuilder_.setMessage(builderForValue.build());
        }

        return this;
      }
      /**
       * <code>optional .google.genomics.v1.Range query_range = 1;</code>
       *
       * <pre>
       * The range of the region in the query genome assembly.
       * </pre>
       */
      public Builder mergeQueryRange(com.google.genomics.v1.Range value) {
        if (queryRangeBuilder_ == null) {
          if (queryRange_ != null) {
            queryRange_ =
              com.google.genomics.v1.Range.newBuilder(queryRange_).mergeFrom(value).buildPartial();
          } else {
            queryRange_ = value;
          }
          onChanged();
        } else {
          queryRangeBuilder_.mergeFrom(value);
        }

        return this;
      }
      /**
       * <code>optional .google.genomics.v1.Range query_range = 1;</code>
       *
       * <pre>
       * The range of the region in the query genome assembly.
       * </pre>
       */
      public Builder clearQueryRange() {
        if (queryRangeBuilder_ == null) {
          queryRange_ = null;
          onChanged();
        } else {
          queryRange_ = null;
          queryRangeBuilder_ = null;
        }

        return this;
      }
      /**
       * <code>optional .google.genomics.v1.Range query_range = 1;</code>
       *
       * <pre>
       * The range of the region in the query genome assembly.
       * </pre>
       */
      public com.google.genomics.v1.Range.Builder getQueryRangeBuilder() {
        
        onChanged();
        return getQueryRangeFieldBuilder().getBuilder();
      }
      /**
       * <code>optional .google.genomics.v1.Range query_range = 1;</code>
       *
       * <pre>
       * The range of the region in the query genome assembly.
       * </pre>
       */
      public com.google.genomics.v1.RangeOrBuilder getQueryRangeOrBuilder() {
        if (queryRangeBuilder_ != null) {
          return queryRangeBuilder_.getMessageOrBuilder();
        } else {
          return queryRange_ == null ?
              com.google.genomics.v1.Range.getDefaultInstance() : queryRange_;
        }
      }
      /**
       * <code>optional .google.genomics.v1.Range query_range = 1;</code>
       *
       * <pre>
       * The range of the region in the query genome assembly.
       * </pre>
       */
      private com.google.protobuf.SingleFieldBuilder<
          com.google.genomics.v1.Range, com.google.genomics.v1.Range.Builder, com.google.genomics.v1.RangeOrBuilder> 
          getQueryRangeFieldBuilder() {
        if (queryRangeBuilder_ == null) {
          queryRangeBuilder_ = new com.google.protobuf.SingleFieldBuilder<
              com.google.genomics.v1.Range, com.google.genomics.v1.Range.Builder, com.google.genomics.v1.RangeOrBuilder>(
                  getQueryRange(),
                  getParentForChildren(),
                  isClean());
          queryRange_ = null;
        }
        return queryRangeBuilder_;
      }

      private com.google.genomics.v1.Range targetRange_ = null;
      private com.google.protobuf.SingleFieldBuilder<
          com.google.genomics.v1.Range, com.google.genomics.v1.Range.Builder, com.google.genomics.v1.RangeOrBuilder> targetRangeBuilder_;
      /**
       * <code>optional .google.genomics.v1.Range target_range = 2;</code>
       *
       * <pre>
       * The range of the region in the target genome assembly.
       * </pre>
       */
      public boolean hasTargetRange() {
        return targetRangeBuilder_ != null || targetRange_ != null;
      }
      /**
       * <code>optional .google.genomics.v1.Range target_range = 2;</code>
       *
       * <pre>
       * The range of the region in the target genome assembly.
       * </pre>
       */
      public com.google.genomics.v1.Range getTargetRange() {
        if (targetRangeBuilder_ == null) {
          return targetRange_ == null ? com.google.genomics.v1.Range.getDefaultInstance() : targetRange_;
        } else {
          return targetRangeBuilder_.getMessage();
        }
      }
      /**
       * <code>optional .google.genomics.v1.Range target_range = 2;</code>
       *
       * <pre>
       * The range of the region in the target genome assembly.
       * </pre>
       */
      public Builder setTargetRange(com.google.genomics.v1.Range value) {
        if (targetRangeBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          targetRange_ = value;
          onChanged();
        } else {
          targetRangeBuilder_.setMessage(value);
        }

        return this;
      }
      /**
       * <code>optional .google.genomics.v1.Range target_range = 2;</code>
       *
       * <pre>
       * The range of the region in the target genome assembly.
       * </pre>
       */
      public Builder setTargetRange(
          com.google.genomics.v1.Range.Builder builderForValue) {
        if (targetRangeBuilder_ == null) {
          targetRange_ = builderForValue.build();
          onChanged();
        } else {
          targetRangeBuilder_.setMessage(builderForValue.build());
        }

        return this;
      }
      /**
       * <code>optional .google.genomics.v1.Range target_range = 2;</code>
       *
       * <pre>
       * The range of the region in the target genome assembly.
       * </pre>
       */
      public Builder mergeTargetRange(com.google.genomics.v1.Range value) {
        if (targetRangeBuilder_ == null) {
          if (targetRange_ != null) {
            targetRange_ =
              com.google.genomics.v1.Range.newBuilder(targetRange_).mergeFrom(value).buildPartial();
          } else {
            targetRange_ = value;
          }
          onChanged();
        } else {
          targetRangeBuilder_.mergeFrom(value);
        }

        return this;
      }
      /**
       * <code>optional .google.genomics.v1.Range target_range = 2;</code>
       *
       * <pre>
       * The range of the region in the target genome assembly.
       * </pre>
       */
      public Builder clearTargetRange() {
        if (targetRangeBuilder_ == null) {
          targetRange_ = null;
          onChanged();
        } else {
          targetRange_ = null;
          targetRangeBuilder_ = null;
        }

        return this;
      }
      /**
       * <code>optional .google.genomics.v1.Range target_range = 2;</code>
       *
       * <pre>
       * The range of the region in the target genome assembly.
       * </pre>
       */
      public com.google.genomics.v1.Range.Builder getTargetRangeBuilder() {
        
        onChanged();
        return getTargetRangeFieldBuilder().getBuilder();
      }
      /**
       * <code>optional .google.genomics.v1.Range target_range = 2;</code>
       *
       * <pre>
       * The range of the region in the target genome assembly.
       * </pre>
       */
      public com.google.genomics.v1.RangeOrBuilder getTargetRangeOrBuilder() {
        if (targetRangeBuilder_ != null) {
          return targetRangeBuilder_.getMessageOrBuilder();
        } else {
          return targetRange_ == null ?
              com.google.genomics.v1.Range.getDefaultInstance() : targetRange_;
        }
      }
      /**
       * <code>optional .google.genomics.v1.Range target_range = 2;</code>
       *
       * <pre>
       * The range of the region in the target genome assembly.
       * </pre>
       */
      private com.google.protobuf.SingleFieldBuilder<
          com.google.genomics.v1.Range, com.google.genomics.v1.Range.Builder, com.google.genomics.v1.RangeOrBuilder> 
          getTargetRangeFieldBuilder() {
        if (targetRangeBuilder_ == null) {
          targetRangeBuilder_ = new com.google.protobuf.SingleFieldBuilder<
              com.google.genomics.v1.Range, com.google.genomics.v1.Range.Builder, com.google.genomics.v1.RangeOrBuilder>(
                  getTargetRange(),
                  getParentForChildren(),
                  isClean());
          targetRange_ = null;
        }
        return targetRangeBuilder_;
      }

      private int targetStrand_ = 0;
      /**
       * <code>optional .com.verily.genomewarp.HomologousRange.TargetStrand target_strand = 3;</code>
       */
      public int getTargetStrandValue() {
        return targetStrand_;
      }
      /**
       * <code>optional .com.verily.genomewarp.HomologousRange.TargetStrand target_strand = 3;</code>
       */
      public Builder setTargetStrandValue(int value) {
        targetStrand_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional .com.verily.genomewarp.HomologousRange.TargetStrand target_strand = 3;</code>
       */
      public com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand getTargetStrand() {
        com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand result = com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand.valueOf(targetStrand_);
        return result == null ? com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand.UNRECOGNIZED : result;
      }
      /**
       * <code>optional .com.verily.genomewarp.HomologousRange.TargetStrand target_strand = 3;</code>
       */
      public Builder setTargetStrand(com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.TargetStrand value) {
        if (value == null) {
          throw new NullPointerException();
        }
        
        targetStrand_ = value.getNumber();
        onChanged();
        return this;
      }
      /**
       * <code>optional .com.verily.genomewarp.HomologousRange.TargetStrand target_strand = 3;</code>
       */
      public Builder clearTargetStrand() {
        
        targetStrand_ = 0;
        onChanged();
        return this;
      }

      private int regionType_ = 0;
      /**
       * <code>optional .com.verily.genomewarp.HomologousRange.RegionType region_type = 4;</code>
       */
      public int getRegionTypeValue() {
        return regionType_;
      }
      /**
       * <code>optional .com.verily.genomewarp.HomologousRange.RegionType region_type = 4;</code>
       */
      public Builder setRegionTypeValue(int value) {
        regionType_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional .com.verily.genomewarp.HomologousRange.RegionType region_type = 4;</code>
       */
      public com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType getRegionType() {
        com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType result = com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType.valueOf(regionType_);
        return result == null ? com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType.UNRECOGNIZED : result;
      }
      /**
       * <code>optional .com.verily.genomewarp.HomologousRange.RegionType region_type = 4;</code>
       */
      public Builder setRegionType(com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange.RegionType value) {
        if (value == null) {
          throw new NullPointerException();
        }
        
        regionType_ = value.getNumber();
        onChanged();
        return this;
      }
      /**
       * <code>optional .com.verily.genomewarp.HomologousRange.RegionType region_type = 4;</code>
       */
      public Builder clearRegionType() {
        
        regionType_ = 0;
        onChanged();
        return this;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:com.verily.genomewarp.HomologousRange)
    }

    // @@protoc_insertion_point(class_scope:com.verily.genomewarp.HomologousRange)
    private static final com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange();
    }

    public static com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<HomologousRange>
        PARSER = new com.google.protobuf.AbstractParser<HomologousRange>() {
      public HomologousRange parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new HomologousRange(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<HomologousRange> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<HomologousRange> getParserForType() {
      return PARSER;
    }

    public com.verily.genomewarp.HomologousRangeOuterClass.HomologousRange getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public interface TransformedVariantsOrBuilder extends
      // @@protoc_insertion_point(interface_extends:com.verily.genomewarp.TransformedVariants)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>optional .com.verily.genomewarp.TransformedVariants.TransformationStatus status = 1;</code>
     */
    int getStatusValue();
    /**
     * <code>optional .com.verily.genomewarp.TransformedVariants.TransformationStatus status = 1;</code>
     */
    com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus getStatus();

    /**
     * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
     *
     * <pre>
     * The list of transformed variants.
     * </pre>
     */
    java.util.List<com.google.genomics.v1.Variant> 
        getVariantsList();
    /**
     * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
     *
     * <pre>
     * The list of transformed variants.
     * </pre>
     */
    com.google.genomics.v1.Variant getVariants(int index);
    /**
     * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
     *
     * <pre>
     * The list of transformed variants.
     * </pre>
     */
    int getVariantsCount();
    /**
     * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
     *
     * <pre>
     * The list of transformed variants.
     * </pre>
     */
    java.util.List<? extends com.google.genomics.v1.VariantOrBuilder> 
        getVariantsOrBuilderList();
    /**
     * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
     *
     * <pre>
     * The list of transformed variants.
     * </pre>
     */
    com.google.genomics.v1.VariantOrBuilder getVariantsOrBuilder(
        int index);
  }
  /**
   * Protobuf type {@code com.verily.genomewarp.TransformedVariants}
   *
   * <pre>
   * Message that contains Variants on the target genome assembly and a status
   * indicating whether the variants were successfully transformed or not.
   * Next ID: 3
   * </pre>
   */
  public  static final class TransformedVariants extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:com.verily.genomewarp.TransformedVariants)
      TransformedVariantsOrBuilder {
    // Use TransformedVariants.newBuilder() to construct.
    private TransformedVariants(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private TransformedVariants() {
      status_ = 0;
      variants_ = java.util.Collections.emptyList();
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private TransformedVariants(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 8: {
              int rawValue = input.readEnum();

              status_ = rawValue;
              break;
            }
            case 18: {
              if (!((mutable_bitField0_ & 0x00000002) == 0x00000002)) {
                variants_ = new java.util.ArrayList<com.google.genomics.v1.Variant>();
                mutable_bitField0_ |= 0x00000002;
              }
              variants_.add(input.readMessage(com.google.genomics.v1.Variant.parser(), extensionRegistry));
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        if (((mutable_bitField0_ & 0x00000002) == 0x00000002)) {
          variants_ = java.util.Collections.unmodifiableList(variants_);
        }
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.verily.genomewarp.HomologousRangeOuterClass.internal_static_com_verily_genomewarp_TransformedVariants_descriptor;
    }

    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.verily.genomewarp.HomologousRangeOuterClass.internal_static_com_verily_genomewarp_TransformedVariants_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.class, com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.Builder.class);
    }

    /**
     * Protobuf enum {@code com.verily.genomewarp.TransformedVariants.TransformationStatus}
     *
     * <pre>
     * The status of the transformation attempt.
     * </pre>
     */
    public enum TransformationStatus
        implements com.google.protobuf.ProtocolMessageEnum {
      /**
       * <code>UNKNOWN_TRANSFORMATION_STATUS = 0;</code>
       */
      UNKNOWN_TRANSFORMATION_STATUS(0, 0),
      /**
       * <code>SUCCESS = 1;</code>
       *
       * <pre>
       * Successful transformation to the target assembly
       * </pre>
       */
      SUCCESS(1, 1),
      /**
       * <code>FAILURE = 2;</code>
       *
       * <pre>
       * Unsuccessful transformation
       * </pre>
       */
      FAILURE(2, 2),
      UNRECOGNIZED(-1, -1),
      ;

      /**
       * <code>UNKNOWN_TRANSFORMATION_STATUS = 0;</code>
       */
      public static final int UNKNOWN_TRANSFORMATION_STATUS_VALUE = 0;
      /**
       * <code>SUCCESS = 1;</code>
       *
       * <pre>
       * Successful transformation to the target assembly
       * </pre>
       */
      public static final int SUCCESS_VALUE = 1;
      /**
       * <code>FAILURE = 2;</code>
       *
       * <pre>
       * Unsuccessful transformation
       * </pre>
       */
      public static final int FAILURE_VALUE = 2;


      public final int getNumber() {
        if (index == -1) {
          throw new java.lang.IllegalArgumentException(
              "Can't get the number of an unknown enum value.");
        }
        return value;
      }

      public static TransformationStatus valueOf(int value) {
        switch (value) {
          case 0: return UNKNOWN_TRANSFORMATION_STATUS;
          case 1: return SUCCESS;
          case 2: return FAILURE;
          default: return null;
        }
      }

      public static com.google.protobuf.Internal.EnumLiteMap<TransformationStatus>
          internalGetValueMap() {
        return internalValueMap;
      }
      private static final com.google.protobuf.Internal.EnumLiteMap<
          TransformationStatus> internalValueMap =
            new com.google.protobuf.Internal.EnumLiteMap<TransformationStatus>() {
              public TransformationStatus findValueByNumber(int number) {
                return TransformationStatus.valueOf(number);
              }
            };

      public final com.google.protobuf.Descriptors.EnumValueDescriptor
          getValueDescriptor() {
        return getDescriptor().getValues().get(index);
      }
      public final com.google.protobuf.Descriptors.EnumDescriptor
          getDescriptorForType() {
        return getDescriptor();
      }
      public static final com.google.protobuf.Descriptors.EnumDescriptor
          getDescriptor() {
        return com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.getDescriptor().getEnumTypes().get(0);
      }

      private static final TransformationStatus[] VALUES = values();

      public static TransformationStatus valueOf(
          com.google.protobuf.Descriptors.EnumValueDescriptor desc) {
        if (desc.getType() != getDescriptor()) {
          throw new java.lang.IllegalArgumentException(
            "EnumValueDescriptor is not for this type.");
        }
        if (desc.getIndex() == -1) {
          return UNRECOGNIZED;
        }
        return VALUES[desc.getIndex()];
      }

      private final int index;
      private final int value;

      private TransformationStatus(int index, int value) {
        this.index = index;
        this.value = value;
      }

      // @@protoc_insertion_point(enum_scope:com.verily.genomewarp.TransformedVariants.TransformationStatus)
    }

    private int bitField0_;
    public static final int STATUS_FIELD_NUMBER = 1;
    private int status_;
    /**
     * <code>optional .com.verily.genomewarp.TransformedVariants.TransformationStatus status = 1;</code>
     */
    public int getStatusValue() {
      return status_;
    }
    /**
     * <code>optional .com.verily.genomewarp.TransformedVariants.TransformationStatus status = 1;</code>
     */
    public com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus getStatus() {
      com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus result = com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus.valueOf(status_);
      return result == null ? com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus.UNRECOGNIZED : result;
    }

    public static final int VARIANTS_FIELD_NUMBER = 2;
    private java.util.List<com.google.genomics.v1.Variant> variants_;
    /**
     * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
     *
     * <pre>
     * The list of transformed variants.
     * </pre>
     */
    public java.util.List<com.google.genomics.v1.Variant> getVariantsList() {
      return variants_;
    }
    /**
     * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
     *
     * <pre>
     * The list of transformed variants.
     * </pre>
     */
    public java.util.List<? extends com.google.genomics.v1.VariantOrBuilder> 
        getVariantsOrBuilderList() {
      return variants_;
    }
    /**
     * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
     *
     * <pre>
     * The list of transformed variants.
     * </pre>
     */
    public int getVariantsCount() {
      return variants_.size();
    }
    /**
     * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
     *
     * <pre>
     * The list of transformed variants.
     * </pre>
     */
    public com.google.genomics.v1.Variant getVariants(int index) {
      return variants_.get(index);
    }
    /**
     * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
     *
     * <pre>
     * The list of transformed variants.
     * </pre>
     */
    public com.google.genomics.v1.VariantOrBuilder getVariantsOrBuilder(
        int index) {
      return variants_.get(index);
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      if (status_ != com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus.UNKNOWN_TRANSFORMATION_STATUS.getNumber()) {
        output.writeEnum(1, status_);
      }
      for (int i = 0; i < variants_.size(); i++) {
        output.writeMessage(2, variants_.get(i));
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (status_ != com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus.UNKNOWN_TRANSFORMATION_STATUS.getNumber()) {
        size += com.google.protobuf.CodedOutputStream
          .computeEnumSize(1, status_);
      }
      for (int i = 0; i < variants_.size(); i++) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(2, variants_.get(i));
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code com.verily.genomewarp.TransformedVariants}
     *
     * <pre>
     * Message that contains Variants on the target genome assembly and a status
     * indicating whether the variants were successfully transformed or not.
     * Next ID: 3
     * </pre>
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:com.verily.genomewarp.TransformedVariants)
        com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariantsOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return com.verily.genomewarp.HomologousRangeOuterClass.internal_static_com_verily_genomewarp_TransformedVariants_descriptor;
      }

      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return com.verily.genomewarp.HomologousRangeOuterClass.internal_static_com_verily_genomewarp_TransformedVariants_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.class, com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.Builder.class);
      }

      // Construct using com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
          getVariantsFieldBuilder();
        }
      }
      public Builder clear() {
        super.clear();
        status_ = 0;

        if (variantsBuilder_ == null) {
          variants_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00000002);
        } else {
          variantsBuilder_.clear();
        }
        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return com.verily.genomewarp.HomologousRangeOuterClass.internal_static_com_verily_genomewarp_TransformedVariants_descriptor;
      }

      public com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants getDefaultInstanceForType() {
        return com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.getDefaultInstance();
      }

      public com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants build() {
        com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants buildPartial() {
        com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants result = new com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants(this);
        int from_bitField0_ = bitField0_;
        int to_bitField0_ = 0;
        result.status_ = status_;
        if (variantsBuilder_ == null) {
          if (((bitField0_ & 0x00000002) == 0x00000002)) {
            variants_ = java.util.Collections.unmodifiableList(variants_);
            bitField0_ = (bitField0_ & ~0x00000002);
          }
          result.variants_ = variants_;
        } else {
          result.variants_ = variantsBuilder_.build();
        }
        result.bitField0_ = to_bitField0_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants) {
          return mergeFrom((com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants other) {
        if (other == com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.getDefaultInstance()) return this;
        if (other.status_ != 0) {
          setStatusValue(other.getStatusValue());
        }
        if (variantsBuilder_ == null) {
          if (!other.variants_.isEmpty()) {
            if (variants_.isEmpty()) {
              variants_ = other.variants_;
              bitField0_ = (bitField0_ & ~0x00000002);
            } else {
              ensureVariantsIsMutable();
              variants_.addAll(other.variants_);
            }
            onChanged();
          }
        } else {
          if (!other.variants_.isEmpty()) {
            if (variantsBuilder_.isEmpty()) {
              variantsBuilder_.dispose();
              variantsBuilder_ = null;
              variants_ = other.variants_;
              bitField0_ = (bitField0_ & ~0x00000002);
              variantsBuilder_ = 
                com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders ?
                   getVariantsFieldBuilder() : null;
            } else {
              variantsBuilder_.addAllMessages(other.variants_);
            }
          }
        }
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }
      private int bitField0_;

      private int status_ = 0;
      /**
       * <code>optional .com.verily.genomewarp.TransformedVariants.TransformationStatus status = 1;</code>
       */
      public int getStatusValue() {
        return status_;
      }
      /**
       * <code>optional .com.verily.genomewarp.TransformedVariants.TransformationStatus status = 1;</code>
       */
      public Builder setStatusValue(int value) {
        status_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional .com.verily.genomewarp.TransformedVariants.TransformationStatus status = 1;</code>
       */
      public com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus getStatus() {
        com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus result = com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus.valueOf(status_);
        return result == null ? com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus.UNRECOGNIZED : result;
      }
      /**
       * <code>optional .com.verily.genomewarp.TransformedVariants.TransformationStatus status = 1;</code>
       */
      public Builder setStatus(com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants.TransformationStatus value) {
        if (value == null) {
          throw new NullPointerException();
        }
        
        status_ = value.getNumber();
        onChanged();
        return this;
      }
      /**
       * <code>optional .com.verily.genomewarp.TransformedVariants.TransformationStatus status = 1;</code>
       */
      public Builder clearStatus() {
        
        status_ = 0;
        onChanged();
        return this;
      }

      private java.util.List<com.google.genomics.v1.Variant> variants_ =
        java.util.Collections.emptyList();
      private void ensureVariantsIsMutable() {
        if (!((bitField0_ & 0x00000002) == 0x00000002)) {
          variants_ = new java.util.ArrayList<com.google.genomics.v1.Variant>(variants_);
          bitField0_ |= 0x00000002;
         }
      }

      private com.google.protobuf.RepeatedFieldBuilder<
          com.google.genomics.v1.Variant, com.google.genomics.v1.Variant.Builder, com.google.genomics.v1.VariantOrBuilder> variantsBuilder_;

      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public java.util.List<com.google.genomics.v1.Variant> getVariantsList() {
        if (variantsBuilder_ == null) {
          return java.util.Collections.unmodifiableList(variants_);
        } else {
          return variantsBuilder_.getMessageList();
        }
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public int getVariantsCount() {
        if (variantsBuilder_ == null) {
          return variants_.size();
        } else {
          return variantsBuilder_.getCount();
        }
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public com.google.genomics.v1.Variant getVariants(int index) {
        if (variantsBuilder_ == null) {
          return variants_.get(index);
        } else {
          return variantsBuilder_.getMessage(index);
        }
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public Builder setVariants(
          int index, com.google.genomics.v1.Variant value) {
        if (variantsBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureVariantsIsMutable();
          variants_.set(index, value);
          onChanged();
        } else {
          variantsBuilder_.setMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public Builder setVariants(
          int index, com.google.genomics.v1.Variant.Builder builderForValue) {
        if (variantsBuilder_ == null) {
          ensureVariantsIsMutable();
          variants_.set(index, builderForValue.build());
          onChanged();
        } else {
          variantsBuilder_.setMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public Builder addVariants(com.google.genomics.v1.Variant value) {
        if (variantsBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureVariantsIsMutable();
          variants_.add(value);
          onChanged();
        } else {
          variantsBuilder_.addMessage(value);
        }
        return this;
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public Builder addVariants(
          int index, com.google.genomics.v1.Variant value) {
        if (variantsBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureVariantsIsMutable();
          variants_.add(index, value);
          onChanged();
        } else {
          variantsBuilder_.addMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public Builder addVariants(
          com.google.genomics.v1.Variant.Builder builderForValue) {
        if (variantsBuilder_ == null) {
          ensureVariantsIsMutable();
          variants_.add(builderForValue.build());
          onChanged();
        } else {
          variantsBuilder_.addMessage(builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public Builder addVariants(
          int index, com.google.genomics.v1.Variant.Builder builderForValue) {
        if (variantsBuilder_ == null) {
          ensureVariantsIsMutable();
          variants_.add(index, builderForValue.build());
          onChanged();
        } else {
          variantsBuilder_.addMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public Builder addAllVariants(
          java.lang.Iterable<? extends com.google.genomics.v1.Variant> values) {
        if (variantsBuilder_ == null) {
          ensureVariantsIsMutable();
          com.google.protobuf.AbstractMessageLite.Builder.addAll(
              values, variants_);
          onChanged();
        } else {
          variantsBuilder_.addAllMessages(values);
        }
        return this;
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public Builder clearVariants() {
        if (variantsBuilder_ == null) {
          variants_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00000002);
          onChanged();
        } else {
          variantsBuilder_.clear();
        }
        return this;
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public Builder removeVariants(int index) {
        if (variantsBuilder_ == null) {
          ensureVariantsIsMutable();
          variants_.remove(index);
          onChanged();
        } else {
          variantsBuilder_.remove(index);
        }
        return this;
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public com.google.genomics.v1.Variant.Builder getVariantsBuilder(
          int index) {
        return getVariantsFieldBuilder().getBuilder(index);
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public com.google.genomics.v1.VariantOrBuilder getVariantsOrBuilder(
          int index) {
        if (variantsBuilder_ == null) {
          return variants_.get(index);  } else {
          return variantsBuilder_.getMessageOrBuilder(index);
        }
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public java.util.List<? extends com.google.genomics.v1.VariantOrBuilder> 
           getVariantsOrBuilderList() {
        if (variantsBuilder_ != null) {
          return variantsBuilder_.getMessageOrBuilderList();
        } else {
          return java.util.Collections.unmodifiableList(variants_);
        }
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public com.google.genomics.v1.Variant.Builder addVariantsBuilder() {
        return getVariantsFieldBuilder().addBuilder(
            com.google.genomics.v1.Variant.getDefaultInstance());
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public com.google.genomics.v1.Variant.Builder addVariantsBuilder(
          int index) {
        return getVariantsFieldBuilder().addBuilder(
            index, com.google.genomics.v1.Variant.getDefaultInstance());
      }
      /**
       * <code>repeated .google.genomics.v1.Variant variants = 2;</code>
       *
       * <pre>
       * The list of transformed variants.
       * </pre>
       */
      public java.util.List<com.google.genomics.v1.Variant.Builder> 
           getVariantsBuilderList() {
        return getVariantsFieldBuilder().getBuilderList();
      }
      private com.google.protobuf.RepeatedFieldBuilder<
          com.google.genomics.v1.Variant, com.google.genomics.v1.Variant.Builder, com.google.genomics.v1.VariantOrBuilder> 
          getVariantsFieldBuilder() {
        if (variantsBuilder_ == null) {
          variantsBuilder_ = new com.google.protobuf.RepeatedFieldBuilder<
              com.google.genomics.v1.Variant, com.google.genomics.v1.Variant.Builder, com.google.genomics.v1.VariantOrBuilder>(
                  variants_,
                  ((bitField0_ & 0x00000002) == 0x00000002),
                  getParentForChildren(),
                  isClean());
          variants_ = null;
        }
        return variantsBuilder_;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:com.verily.genomewarp.TransformedVariants)
    }

    // @@protoc_insertion_point(class_scope:com.verily.genomewarp.TransformedVariants)
    private static final com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants();
    }

    public static com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<TransformedVariants>
        PARSER = new com.google.protobuf.AbstractParser<TransformedVariants>() {
      public TransformedVariants parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new TransformedVariants(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<TransformedVariants> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<TransformedVariants> getParserForType() {
      return PARSER;
    }

    public com.verily.genomewarp.HomologousRangeOuterClass.TransformedVariants getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_com_verily_genomewarp_HomologousRange_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_com_verily_genomewarp_HomologousRange_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_com_verily_genomewarp_TransformedVariants_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_com_verily_genomewarp_TransformedVariants_fieldAccessorTable;

  public static com.google.protobuf.Descriptors.FileDescriptor
      getDescriptor() {
    return descriptor;
  }
  private static com.google.protobuf.Descriptors.FileDescriptor
      descriptor;
  static {
    java.lang.String[] descriptorData = {
      "\n\026homologous_range.proto\022\025com.verily.gen" +
      "omewarp\032\036google/genomics/v1/range.proto\032" +
      "!google/genomics/v1/variants.proto\"\277\003\n\017H" +
      "omologousRange\022.\n\013query_range\030\001 \001(\0132\031.go" +
      "ogle.genomics.v1.Range\022/\n\014target_range\030\002" +
      " \001(\0132\031.google.genomics.v1.Range\022J\n\rtarge" +
      "t_strand\030\003 \001(\01623.com.verily.genomewarp.H" +
      "omologousRange.TargetStrand\022F\n\013region_ty" +
      "pe\030\004 \001(\01621.com.verily.genomewarp.Homolog" +
      "ousRange.RegionType\"S\n\014TargetStrand\022\031\n\025U",
      "NKNOWN_TARGET_STRAND\020\000\022\023\n\017POSITIVE_STRAN" +
      "D\020\001\022\023\n\017NEGATIVE_STRAND\020\002\"b\n\nRegionType\022\027" +
      "\n\023UNKNOWN_REGION_TYPE\020\000\022\r\n\tIDENTICAL\020\001\022\024" +
      "\n\020MISMATCHED_BASES\020\004\022\026\n\022ALIGNMENT_REQUIR" +
      "ED\020\003\"\352\001\n\023TransformedVariants\022O\n\006status\030\001" +
      " \001(\0162?.com.verily.genomewarp.Transformed" +
      "Variants.TransformationStatus\022-\n\010variant" +
      "s\030\002 \003(\0132\033.google.genomics.v1.Variant\"S\n\024" +
      "TransformationStatus\022!\n\035UNKNOWN_TRANSFOR" +
      "MATION_STATUS\020\000\022\013\n\007SUCCESS\020\001\022\013\n\007FAILURE\020",
      "\002B\027\n\025com.verily.genomewarpb\006proto3"
    };
    com.google.protobuf.Descriptors.FileDescriptor.InternalDescriptorAssigner assigner =
        new com.google.protobuf.Descriptors.FileDescriptor.    InternalDescriptorAssigner() {
          public com.google.protobuf.ExtensionRegistry assignDescriptors(
              com.google.protobuf.Descriptors.FileDescriptor root) {
            descriptor = root;
            return null;
          }
        };
    com.google.protobuf.Descriptors.FileDescriptor
      .internalBuildGeneratedFileFrom(descriptorData,
        new com.google.protobuf.Descriptors.FileDescriptor[] {
          com.google.genomics.v1.RangeProto.getDescriptor(),
          com.google.genomics.v1.VariantsProto.getDescriptor(),
        }, assigner);
    internal_static_com_verily_genomewarp_HomologousRange_descriptor =
      getDescriptor().getMessageTypes().get(0);
    internal_static_com_verily_genomewarp_HomologousRange_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_com_verily_genomewarp_HomologousRange_descriptor,
        new java.lang.String[] { "QueryRange", "TargetRange", "TargetStrand", "RegionType", });
    internal_static_com_verily_genomewarp_TransformedVariants_descriptor =
      getDescriptor().getMessageTypes().get(1);
    internal_static_com_verily_genomewarp_TransformedVariants_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_com_verily_genomewarp_TransformedVariants_descriptor,
        new java.lang.String[] { "Status", "Variants", });
    com.google.genomics.v1.RangeProto.getDescriptor();
    com.google.genomics.v1.VariantsProto.getDescriptor();
  }

  // @@protoc_insertion_point(outer_class_scope)
}
